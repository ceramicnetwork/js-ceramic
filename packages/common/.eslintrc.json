{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint", "import"],
  "rules": {
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/ban-ts-ignore": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "import/no-unresolved": "error",
    "import/default": "off"
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      "typescript": {
        "extensions": []
      },
      "node": true
    }
  },
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/typescript"
  ]
}
