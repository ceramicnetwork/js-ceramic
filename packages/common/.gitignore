node_modules/
coverage/
lib/
src/generated/
