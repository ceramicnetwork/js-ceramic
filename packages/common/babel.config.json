{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "node": "16"
        },
        "useBuiltIns": false,
        "debug": false
      }
    ],
    "@babel/preset-typescript"
  ],
  "plugins": [
    [
      "@babel/plugin-transform-modules-commonjs",
      {
        "allowTopLevelThis": true
      }
    ],
    ["@babel/plugin-proposal-decorators", { "legacy": true }]
  ]
}
