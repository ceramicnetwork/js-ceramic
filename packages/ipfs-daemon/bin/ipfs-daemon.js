#!/usr/bin/env node

await import('../lib/bin/ipfs-daemon.js')
