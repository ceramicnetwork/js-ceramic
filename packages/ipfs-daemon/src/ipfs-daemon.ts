import { createController } from './create-ipfs.js'
import * as dagJose from 'dag-jose'
import { IpfsTopology } from '@ceramicnetwork/ipfs-topology'
import { DiagnosticsLogger, LogLevel } from '@ceramicnetwork/common'
import { HealthcheckServer } from './healthcheck-server.js'
import path from 'path'
import os from 'os'
import * as Ctl from 'ipfsd-ctl'

export interface Configuration {
  tcpHost: string
  ipfsPath: string
  announceAddressList: string[]
  ipfsCreateIfMissing: boolean
  ipfsSwarmTcpPort: number
  ipfsSwarmWsPort: number
  ipfsApiPort: number
  ipfsGatewayPort: number
  ipfsEnableGateway: boolean
  ipfsDhtServerMode: boolean
  ipfsEnablePubsub: boolean
  ipfsPubsubTopics: string[]
  ipfsBootstrap: string[]
  ceramicNetwork: string
  useCentralizedPeerDiscovery: boolean
  healthcheckEnabled: boolean
  healthcheckPort: number
  logger: DiagnosticsLogger
}

function fromBooleanInput(input: string | undefined, byDefault?: boolean): boolean {
  if (input) {
    return input == 'true'
  } else {
    return Boolean(byDefault)
  }
}

export class IpfsDaemon {
  topology?: IpfsTopology
  healthcheck?: HealthcheckServer

  constructor(readonly configuration: Configuration, readonly ipfsd: Ctl.Controller) {}

  static async create(props: Partial<Configuration> = {}): Promise<IpfsDaemon> {
    const ceramicNetwork = props.ceramicNetwork || process.env.CERAMIC_NETWORK
    const useCentralizedPeerDiscovery =
      props.useCentralizedPeerDiscovery ?? process.env.NODE_ENV != 'test'
    const ipfsBootstrap =
      props.ipfsBootstrap ||
      (process.env.IPFS_BOOTSTRAP ? process.env.IPFS_BOOTSTRAP.split(' ') : [])
    const ipfsCreateIfMissing =
      props.ipfsCreateIfMissing ?? fromBooleanInput(process.env.IPFS_CREATE_IF_MISSING, true)
    const configuration: Configuration = {
      tcpHost: props.tcpHost || process.env.TCP_HOST || '0.0.0.0',
      ipfsPath: props.ipfsPath || process.env.IPFS_PATH || path.join(os.homedir(), '.goipfs'),
      ipfsCreateIfMissing: ipfsCreateIfMissing,

      announceAddressList:
        props.announceAddressList ?? (process.env.ANNOUNCE_ADDRESS_LIST?.split(',') || []),
      ipfsSwarmTcpPort: props.ipfsSwarmTcpPort || Number(process.env.IPFS_SWARM_TCP_PORT) || 4011,
      ipfsSwarmWsPort: props.ipfsSwarmWsPort || Number(process.env.IPFS_SWARM_WS_PORT) || 4012,

      ipfsApiPort: props.ipfsApiPort || Number(process.env.IPFS_API_PORT) || 5011,

      ipfsEnableGateway:
        props.ipfsEnableGateway ?? fromBooleanInput(process.env.IPFS_ENABLE_GATEWAY, false),
      ipfsGatewayPort: props.ipfsGatewayPort || Number(process.env.IPFS_GATEWAY_PORT) || 9011,
      ipfsDhtServerMode:
        props.ipfsDhtServerMode ?? fromBooleanInput(process.env.IPFS_DHT_SERVER_MODE, false),

      ipfsEnablePubsub:
        props.ipfsEnablePubsub ?? fromBooleanInput(process.env.IPFS_ENABLE_PUBSUB, true),
      ipfsPubsubTopics:
        props.ipfsPubsubTopics ??
        (process.env.IPFS_PUBSUB_TOPICS ? process.env.IPFS_PUBSUB_TOPICS.split(' ') : []),

      ipfsBootstrap: ipfsBootstrap,
      ceramicNetwork: ceramicNetwork,
      useCentralizedPeerDiscovery: useCentralizedPeerDiscovery,
      healthcheckEnabled:
        props.healthcheckEnabled ?? fromBooleanInput(process.env.HEALTHCHECK_ENABLED, false),
      healthcheckPort: props.healthcheckPort || Number(process.env.HEALTHCHECK_PORT) || 8011,
      logger: props.logger ?? new DiagnosticsLogger(LogLevel.important, false),
    }

    const ipfsd = await createController(
      {
        start: false,
        init: {
          allowNew: configuration.ipfsCreateIfMissing,
        },
        repo: configuration.ipfsPath,
        ipld: { codecs: [dagJose] },
        libp2p: {
          config: {
            dht: {
              enabled: false,
              clientMode: !configuration.ipfsDhtServerMode,
            },
            pubsub: {
              enabled: configuration.ipfsEnablePubsub,
            },
          },
          addresses: {
            announce: configuration.announceAddressList,
          },
        },
        preload: {
          enabled: false,
        },
        config: {
          Addresses: {
            Swarm: [
              `/ip4/${configuration.tcpHost}/tcp/${configuration.ipfsSwarmTcpPort}`,
              `/ip4/${configuration.tcpHost}/tcp/${configuration.ipfsSwarmWsPort}/ws`,
            ],
            API: `/ip4/${configuration.tcpHost}/tcp/${configuration.ipfsApiPort}`,
            ...(configuration.ipfsEnableGateway && {
              Gateway: `/ip4/${configuration.tcpHost}/tcp/${configuration.ipfsGatewayPort}`,
            }),
            Announce: configuration.announceAddressList,
          },
          API: {
            HTTPHeaders: {
              'Access-Control-Allow-Origin': ['*'],
              'Access-Control-Allow-Methods': ['GET', 'POST'],
              'Access-Control-Allow-Headers': ['Authorization'],
              'Access-Control-Expose-Headers': ['Location'],
              'Access-Control-Allow-Credentials': ['true'],
            },
          },
          Routing: {
            Type: configuration.ipfsDhtServerMode ? 'dhtserver' : 'dhtclient',
          },
          Bootstrap: configuration.ipfsBootstrap,
        },
      },
      false
    )

    return new IpfsDaemon(configuration, ipfsd)
  }

  async start(): Promise<IpfsDaemon> {
    await this.ipfsd.start()

    this.configuration.logger.imp('IPFS API server listening on ' + this.configuration.ipfsApiPort)

    if (this.configuration.ipfsEnableGateway) {
      this.configuration.logger.imp(
        'IPFS Gateway server listening on ' + this.configuration.ipfsGatewayPort
      )
    }

    if (this.configuration.useCentralizedPeerDiscovery) {
      this.topology = new IpfsTopology(
        this.ipfsd.api,
        this.configuration.ceramicNetwork,
        this.configuration.logger
      )
      await this.topology.start()
    }

    if (this.configuration.healthcheckEnabled) {
      this.healthcheck = new HealthcheckServer(
        this.ipfsd.api,
        this.configuration.healthcheckPort,
        this.configuration.tcpHost,
        this.configuration.logger
      )
      await this.healthcheck.start()
    }

    await Promise.all(
      this.configuration.ipfsPubsubTopics.map((topic) => {
        return this.ipfsd.api.pubsub.subscribe(topic, () => {
          // Do Nothing
        })
      })
    )
    return this
  }

  async stop(): Promise<void> {
    this.topology?.stop()
    await this.ipfsd.stop()
  }
}
