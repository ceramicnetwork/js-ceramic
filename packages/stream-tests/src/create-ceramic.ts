import mergeOpts from 'merge-options'
import { CeramicConfig, Ceramic } from '@ceramicnetwork/core'
import { IpfsApi } from '@ceramicnetwork/common'
import tmp from 'tmp-promise'
import { createDid } from './create_did.js'

export async function createCeramic(
  ipfs: IpfsApi,
  config: CeramicConfig & { seed?: string } = {}
): Promise<Ceramic> {
  const stateStoreDirectory = await tmp.tmpName()
  const appliedConfig: CeramicConfig = mergeOpts(
    {
      stateStoreDirectory: stateStoreDirectory,
      anchorOnRequest: false,
      streamCacheLimit: 100,
      pubsubTopic: '/ceramic/inmemory/test', // necessary so Ceramic instances can talk to each other
      indexing: {
        db: `sqlite://${stateStoreDirectory}/ceramic.sqlite`,
        models: [],
        allowQueriesBeforeHistoricalSync: false,
      },
    },
    config
  )
  const ceramic = await Ceramic.create(ipfs, appliedConfig)
  const did = createDid(ceramic, appliedConfig.seed || 'SEED')
  await ceramic.setDID(did)
  await did.authenticate()

  return ceramic
}
