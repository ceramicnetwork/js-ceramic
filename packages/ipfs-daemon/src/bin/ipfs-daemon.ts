#!/usr/bin/env node

import { IpfsDaemon } from '../ipfs-daemon.js'

process.on('uncaughtException', (err) => {
  console.log(err) // just log for now
})

IpfsDaemon.create()
  .then((ipfs) => ipfs.start())
  .catch((e) => console.error(e))
