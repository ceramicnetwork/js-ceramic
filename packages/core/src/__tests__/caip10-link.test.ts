import { jest } from '@jest/globals'
import tmp from 'tmp-promise'
import { Caip10Link } from '@ceramicnetwork/stream-caip10-link'
import { AnchorStatus, StreamUtils, IpfsApi, TestUtils } from '@ceramicnetwork/common'
import MockDate from 'mockdate'
import type { Ceramic } from '../ceramic.js'
import { createIPFS } from '@ceramicnetwork/ipfs-daemon'
import { createCeramic } from './create-ceramic.js'

const DID_USED = 'did:3:bafysdfwefwe'
const LEGACY_ACCOUNT = '0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1337'
const ACCOUNT = 'eip155:1337:0x8fe2c4516e920425e177658aaac451ca0463ed69'
const FAMILY = 'caip10-eip155:1337'
const PROOF = {
  version: 2,
  type: 'ethereum-eoa',
  message:
    'Link this account to your identity\n' +
    '\n' +
    'did:3:bafysdfwefwe \n' +
    'Timestamp: 1641461695',
  signature:
    '0xbbc62f3184daeb0b5792b50797c493f144b5811a5b634e450329bcdc919f5e2c20166bfde97748fd7ec3281af95230c056e2a89ce9f07420b326440f158d05e21c',
  account: 'eip155:1337:0x8fe2c4516e920425e177658aaac451ca0463ed69',
  timestamp: 1641461695,
}
const LEGACY_PROOF = {
  ...PROOF,
  account: LEGACY_ACCOUNT,
}
const EMPTY_DID_PROOF = {
  version: 2,
  type: 'ethereum-eoa',
  message: 'Link this account to your identity\n\n \nTimestamp: 1641462800',
  signature:
    '0xc353c06cf5d77c5105d72779f1de8a1f38ebead0f7686e9d90e83e445d5a7a640cb50df30cd268ca4dffcf6ffcf71dce7c53d4688e074990d00b5e06058cd87d1b',
  account: 'eip155:1337:0x8fe2c4516e920425e177658aaac451ca0463ed69',
  timestamp: 1641462800,
}

describe('Ceramic API', () => {
  jest.setTimeout(60000)
  let ipfs: IpfsApi

  beforeAll(async () => {
    ipfs = await createIPFS()
    MockDate.set('2018-10-01') // So that the anchors happen at a predictable blockNumber/blockTimestamp
  })

  afterAll(async () => {
    MockDate.reset()
    await ipfs.stop()
  })

  describe('Caip10Link test', () => {
    let ceramic: Ceramic
    let authProvider
    let tmpFolder: tmp.DirectoryResult

    beforeEach(async () => {
      tmpFolder = await tmp.dir({ unsafeCleanup: true })
      ceramic = await createCeramic(ipfs, {
        stateStoreDirectory: tmpFolder.path,
        anchorOnRequest: false,
        verifySignatures: false,
      })
      authProvider = { createLink: jest.fn() }
    })

    afterEach(async () => {
      await ceramic.close()
      await tmpFolder.cleanup()
    })

    it('Create from valid account id', async () => {
      const link = await Caip10Link.fromAccount(ceramic, ACCOUNT)
      expect(link.metadata).toEqual({
        controllers: [LEGACY_ACCOUNT.toLowerCase()],
        family: FAMILY,
      })
      expect(link.did).toBeNull()
      expect(link.state.log).toHaveLength(1)
      expect(link.state).toMatchSnapshot()
    })

    it('Create from legacy account id', async () => {
      const link = await Caip10Link.fromAccount(ceramic, LEGACY_ACCOUNT)
      expect(link.metadata).toEqual({
        controllers: [LEGACY_ACCOUNT.toLowerCase()],
        family: FAMILY,
      })
      expect(link.did).toBeNull()
      expect(link.state.log).toHaveLength(1)
      expect(link.state).toMatchSnapshot()
    })

    it('invalid account id', async () => {
      const invalid1 = '0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8'
      const invalid2 = '0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155'
      const invalid3 = '@eip155:1'
      await expect(Caip10Link.fromAccount(ceramic, invalid1)).rejects.toThrow(
        /Invalid accountId provided/
      )
      await expect(Caip10Link.fromAccount(ceramic, invalid2)).rejects.toThrow(
        /Invalid chainId provided/
      )
      await expect(Caip10Link.fromAccount(ceramic, invalid3)).rejects.toThrow(
        /Invalid accountId provided/
      )
    })

    it('Create and link DID', async () => {
      authProvider.createLink.mockReturnValueOnce(PROOF)

      const link = await Caip10Link.fromAccount(ceramic, LEGACY_ACCOUNT)
      await link.setDid(DID_USED, authProvider, { anchor: false })

      expect(link.did).toEqual(DID_USED)
      expect(link.state.log).toHaveLength(2)
      expect(authProvider.createLink).toHaveBeenCalledTimes(1)
      expect(link.state).toMatchSnapshot()
    })

    it('Create and link DID with legacy CAIP proof', async () => {
      authProvider.createLink.mockReturnValueOnce(LEGACY_PROOF)

      const link = await Caip10Link.fromAccount(ceramic, ACCOUNT)
      await link.setDid(DID_USED, authProvider, { anchor: false })

      expect(link.did).toEqual(DID_USED)
      expect(link.state.log).toHaveLength(2)
      expect(authProvider.createLink).toHaveBeenCalledTimes(1)
      expect(link.state).toMatchSnapshot()
    })

    it('Created with same address loads same doc', async () => {
      authProvider.createLink.mockReturnValueOnce(PROOF)

      const link1 = await Caip10Link.fromAccount(ceramic, LEGACY_ACCOUNT)
      await link1.setDid(DID_USED, authProvider, { anchor: false })

      const link2 = await Caip10Link.fromAccount(ceramic, LEGACY_ACCOUNT)
      expect(link1.id).toEqual(link2.id)
      expect(link1.did).toEqual(DID_USED)
      expect(link2.did).toEqual(link1.did)
      expect(StreamUtils.serializeState(link2.state)).toEqual(
        StreamUtils.serializeState(link1.state)
      )
    })

    it('Load works', async () => {
      authProvider.createLink.mockReturnValueOnce(PROOF)

      const link1 = await Caip10Link.fromAccount(ceramic, LEGACY_ACCOUNT)
      await link1.setDid(DID_USED, authProvider, { anchor: false })

      const link2 = await Caip10Link.load(ceramic, link1.id)
      expect(link1.id).toEqual(link2.id)
      expect(link1.did).toEqual(DID_USED)
      expect(link2.did).toEqual(link1.did)
      expect(StreamUtils.serializeState(link2.state)).toEqual(
        StreamUtils.serializeState(link1.state)
      )
    })

    it('Anchoring works', async () => {
      authProvider.createLink.mockReturnValueOnce(PROOF)

      const link = await Caip10Link.fromAccount(ceramic, LEGACY_ACCOUNT, { anchor: true })
      expect(link.state.anchorStatus).toEqual(AnchorStatus.PENDING)
      await TestUtils.anchorUpdate(ceramic, link)
      expect(link.state.anchorStatus).toEqual(AnchorStatus.ANCHORED)

      await link.setDid(DID_USED, authProvider, { anchor: true })
      expect(link.state.anchorStatus).toEqual(AnchorStatus.PENDING)
      await TestUtils.anchorUpdate(ceramic, link)
      expect(link.state.anchorStatus).toEqual(AnchorStatus.ANCHORED)

      expect(link.did).toEqual(DID_USED)
      expect(link.state.log).toHaveLength(4)
      expect(link.state).toMatchSnapshot()
    })

    it('Throws when linking to invalid DID ', async () => {
      const link = await Caip10Link.fromAccount(ceramic, ACCOUNT)

      const invalidDids = [
        'did:incomplete',
        '  oddPrefixdid:stuff:stuff',
        'did:CAPITALS:stuff',
        'did:&&specialChar:stuff',
        'did:test:%1maw',
      ]

      await Promise.all(
        invalidDids.map(async (did) => {
          await expect(link.setDid(did, authProvider)).rejects.toThrow(/DID is not valid/)
        })
      )
    })

    it('Clear did works', async () => {
      const link = await Caip10Link.fromAccount(ceramic, LEGACY_ACCOUNT)
      await expect(link.did).toBeNull()

      authProvider.createLink.mockReturnValueOnce(PROOF)
      await link.setDid(DID_USED, authProvider)
      await expect(link.did).toEqual(DID_USED)

      authProvider.createLink.mockReturnValueOnce(EMPTY_DID_PROOF)
      await link.clearDid(authProvider)
      await expect(link.did).toBeNull()
    })
  })
})
