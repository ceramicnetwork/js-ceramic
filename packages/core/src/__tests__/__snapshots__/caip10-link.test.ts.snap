// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ceramic API Caip10Link test Anchoring works 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 1538352000,
    "blockTimestamp": 1538352000,
    "chainId": "inmemory:12345",
    "root": Object {
      "code": 113,
      "hash": Uint8Array [
        18,
        32,
        9,
        14,
        51,
        24,
        226,
        153,
        52,
        12,
        23,
        139,
        247,
        206,
        244,
        238,
        83,
        26,
        212,
        233,
        32,
        85,
        142,
        143,
        140,
        100,
        104,
        77,
        49,
        96,
        208,
        39,
        136,
        11,
      ],
      "version": 1,
    },
    "txHash": Object {
      "code": 147,
      "hash": Uint8Array [
        27,
        32,
        24,
        49,
        225,
        228,
        11,
        171,
        49,
        208,
        15,
        26,
        103,
        65,
        213,
        144,
        208,
        238,
        98,
        25,
        53,
        119,
        40,
        180,
        18,
        220,
        137,
        238,
        244,
        57,
        124,
        243,
        123,
        170,
      ],
      "version": 1,
    },
    "version": 1,
  },
  "anchorStatus": 3,
  "content": "did:3:bafysdfwefwe",
  "log": Array [
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          220,
          69,
          11,
          211,
          217,
          11,
          32,
          171,
          217,
          181,
          63,
          193,
          60,
          28,
          62,
          144,
          62,
          101,
          140,
          47,
          33,
          57,
          154,
          254,
          131,
          167,
          84,
          49,
          14,
          178,
          67,
          157,
        ],
        "version": 1,
      },
      "timestamp": 1538352000,
      "type": 0,
    },
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          129,
          245,
          222,
          196,
          76,
          202,
          40,
          189,
          78,
          211,
          167,
          93,
          117,
          43,
          200,
          130,
          141,
          127,
          22,
          19,
          184,
          210,
          163,
          129,
          49,
          64,
          220,
          161,
          251,
          117,
          140,
          143,
        ],
        "version": 1,
      },
      "timestamp": 1538352000,
      "type": 2,
    },
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          9,
          14,
          51,
          24,
          226,
          153,
          52,
          12,
          23,
          139,
          247,
          206,
          244,
          238,
          83,
          26,
          212,
          233,
          32,
          85,
          142,
          143,
          140,
          100,
          104,
          77,
          49,
          96,
          208,
          39,
          136,
          11,
        ],
        "version": 1,
      },
      "timestamp": 1538352000,
      "type": 1,
    },
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          66,
          255,
          10,
          137,
          219,
          36,
          117,
          166,
          34,
          187,
          210,
          23,
          199,
          127,
          126,
          40,
          54,
          123,
          237,
          242,
          136,
          58,
          66,
          224,
          109,
          24,
          123,
          3,
          49,
          116,
          106,
          19,
        ],
        "version": 1,
      },
      "timestamp": 1538352000,
      "type": 2,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1337",
    ],
    "family": "caip10-eip155:1337",
    "lastUpdate": 1641461695,
  },
  "signature": 2,
  "type": 1,
}
`;

exports[`Ceramic API Caip10Link test Create and link DID 1`] = `
Object {
  "anchorStatus": 0,
  "content": null,
  "log": Array [
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          220,
          69,
          11,
          211,
          217,
          11,
          32,
          171,
          217,
          181,
          63,
          193,
          60,
          28,
          62,
          144,
          62,
          101,
          140,
          47,
          33,
          57,
          154,
          254,
          131,
          167,
          84,
          49,
          14,
          178,
          67,
          157,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          79,
          176,
          3,
          183,
          245,
          175,
          148,
          158,
          151,
          111,
          100,
          108,
          165,
          35,
          6,
          223,
          111,
          125,
          124,
          204,
          67,
          32,
          16,
          177,
          79,
          108,
          247,
          92,
          111,
          33,
          168,
          170,
        ],
        "version": 1,
      },
      "type": 1,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1337",
    ],
    "family": "caip10-eip155:1337",
  },
  "next": Object {
    "content": "did:3:bafysdfwefwe",
    "metadata": Object {
      "controllers": Array [
        "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1337",
      ],
      "family": "caip10-eip155:1337",
      "lastUpdate": 1641461695,
    },
  },
  "signature": 2,
  "type": 1,
}
`;

exports[`Ceramic API Caip10Link test Create and link DID with legacy CAIP proof 1`] = `
Object {
  "anchorStatus": 0,
  "content": null,
  "log": Array [
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          220,
          69,
          11,
          211,
          217,
          11,
          32,
          171,
          217,
          181,
          63,
          193,
          60,
          28,
          62,
          144,
          62,
          101,
          140,
          47,
          33,
          57,
          154,
          254,
          131,
          167,
          84,
          49,
          14,
          178,
          67,
          157,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          129,
          61,
          78,
          166,
          67,
          96,
          81,
          131,
          252,
          108,
          54,
          180,
          74,
          78,
          6,
          6,
          46,
          244,
          173,
          168,
          246,
          204,
          9,
          186,
          83,
          90,
          227,
          162,
          129,
          77,
          220,
          239,
        ],
        "version": 1,
      },
      "type": 1,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1337",
    ],
    "family": "caip10-eip155:1337",
  },
  "next": Object {
    "content": "did:3:bafysdfwefwe",
    "metadata": Object {
      "controllers": Array [
        "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1337",
      ],
      "family": "caip10-eip155:1337",
      "lastUpdate": 1641461695,
    },
  },
  "signature": 2,
  "type": 1,
}
`;

exports[`Ceramic API Caip10Link test Create from legacy account id 1`] = `
Object {
  "anchorStatus": 0,
  "content": null,
  "log": Array [
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          220,
          69,
          11,
          211,
          217,
          11,
          32,
          171,
          217,
          181,
          63,
          193,
          60,
          28,
          62,
          144,
          62,
          101,
          140,
          47,
          33,
          57,
          154,
          254,
          131,
          167,
          84,
          49,
          14,
          178,
          67,
          157,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1337",
    ],
    "family": "caip10-eip155:1337",
  },
  "next": Object {
    "content": null,
  },
  "signature": 0,
  "type": 1,
}
`;

exports[`Ceramic API Caip10Link test Create from valid account id 1`] = `
Object {
  "anchorStatus": 0,
  "content": null,
  "log": Array [
    Object {
      "cid": Object {
        "code": 113,
        "hash": Uint8Array [
          18,
          32,
          220,
          69,
          11,
          211,
          217,
          11,
          32,
          171,
          217,
          181,
          63,
          193,
          60,
          28,
          62,
          144,
          62,
          101,
          140,
          47,
          33,
          57,
          154,
          254,
          131,
          167,
          84,
          49,
          14,
          178,
          67,
          157,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "0x8fe2c4516e920425e177658aaac451ca0463ed69@eip155:1337",
    ],
    "family": "caip10-eip155:1337",
  },
  "next": Object {
    "content": null,
  },
  "signature": 0,
  "type": 1,
}
`;
